Refactor: quick clean-up of iteration idioms in the `vfs` crate

Some-Dood opened the pull request:
This PR cleans up some of the iteration idioms used in the `vfs` crate. Most of the changes simply converted `for` loops into their `std::iter::Iterator`-method counterpart. Other changes required some inversion of logic to accommodate for better short-circuiting. Overall, there should be no behavioral changes. If there are any stylistic issues, I will gladly adhere to them and adjust the PR accordingly. Thanks!

matklad commented on crates/vfs/src/lib.rs:
```diff
@@ -137,13 +139,12 @@ impl Vfs {
+            let _ = self.get(file_id).as_ref()?;
+            let path = self.interner.lookup(file_id);
+            Some((file_id, path))
+        })
```
The original formulation is intentional: every `.filter(...).map(...)` sequence can be re-written as `.filter_map`, but converse is not true -- if `filter_map` "changes the types", then it can't be replaced with separate `filter` and `map`. 

That's why, when it is possible to not use `filter_map`, I avoid using it: using less powerful construct makes the code more readable, because there's less things that the code can do.

matklad commented on crates/vfs/src/loader.rs:
```diff
@@ -137,10 +137,7 @@ impl Directories {
-            return false;
-        }
-        self.includes_path(path)
+        self.extensions.iter().any(|it| it.as_str() == ext) && self.includes_path(path)
```
Here, splitting a complex conditional into two parts was intentional. We *can* mechanically merge them into one boolean expression, but it's useful to show that the conditions aer logically independent. Rather than merging them, I'd add a blank like after the `}` in the original code, to signify "we've dealt with file extension, and now lets do a check for path inclusion"

matklad commented on crates/vfs/src/loader.rs:
```diff
@@ -158,25 +155,23 @@ impl Directories {
-                    Some(prev) if prev.starts_with(incl) => prev,
-                    _ => incl,
-                })
+        let include = self.include.iter().fold(None::<&AbsPathBuf>, |include, incl| {
```
Here, the usage of `fold` is forced -- a `for` loop is more natural.

matklad reviewed (commented):
I'd same most of these were written that way intentionally. The `.any` is a good catch though. 

I've tried to spell out rationale for our code style in https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/style.md#functional-combinators

Some-Dood commented:
> I'd say most of these were written that way intentionally. The `.any` is a good catch though.

Thanks for the constructive feedback! I must agree that the usage of `?` does seem like I've used it _just because_. I will definitely revert it back to the original.

Regarding the `filter_map`, I did not realize (until now) how powerful that separation was from a readability standpoint. It had never occurred to me that "using less powerful constructs" expressed the intent of the code more clearly and effectively. Previously, my intention behind the `filter_map` was that it seemed really intuitive—and dare I say, _tempting_—to collapse all `filter` + `map` operations into one big `filter_map`. As you have pointed out, this is not always the best course of action.

> I've tried to spell out rationale for our code style in https://github.com/rust-analyzer/rust-analyzer/blob/master/docs/dev/style.md#functional-combinators

This is definitely a helpful document. Thanks for taking the time to write it up! The part about "bending the code to fit into some combinator" is definitely most relevant in this PR. 😅

Some-Dood commented:
Hello there @matklad! Thanks again for your helpful feedback. I have addressed the stylistic issues you raised. I have also added the [code comments you wished were present](https://github.com/rust-analyzer/rust-analyzer/pull/9836#discussion_r685953381) in `Directories::contains_file`. Please do let me know if there are any other changes that must be made.

matklad commented:
bors r+
